kvstorage: complete RaftReplicaID migration

tbg opened the pull request:
As of v22.1[^1], we always write the RaftReplicaID when creating a
Replica or updating a snapshot. However, since this is
persisted state that could've originated in older versions and not
updated yet, we couldn't rely on a persisted ReplicaID yet.

This commit adds code to the `(*Store).Start` boot sequence that

- persists a RaftReplicaID for all initialized replicas (using the
  ReplicaID from the descriptor)
- deletes all uninitialized replicas lacking RaftReplicaID (since we don't know their
  ReplicaID at this point).

The second item in theory violates Raft invariants, as uninitialized
Replicas are allowed to vote (though they then cannot accept log
entries). So in theory:

- an uninitialized replica casts a decisive vote for a leader
- it restarts
- code in this commit removes the uninited replica (and its vote)
- delayed MsgVote from another leader arrives
- it casts another vote for the same term for a dueling leader
- now there are two leaders in the same term.

The above in addition presupposes that the two leaders cannot
communicate with each other. Also, even if that is the case, since the
two leaders cannot append to the uninitialized replica (it doesn't
accept entries), we also need additional voters to return at the exact
right time.

Since an uninitialized replica without RaftReplicaID in is necessarily
at least one release old, this is exceedingly unlikely and we will
live with this theoretical risk.

This PR also adds a first stab at a datadriven test harness for
`kvstorage` which is likely to be of use for #93247.

[^1]: https://github.com/cockroachdb/cockroach/pull/75761

Epic: [CRDB-220](https://cockroachlabs.atlassian.net/browse/CRDB-220)
Release note: None

tbg committed:
kvstorage: add datadriven harness

See https://github.com/cockroachdb/cockroach/issues/93310. This is
also the beginning of
https://github.com/cockroachdb/cockroach/issues/93247.

Epic: CRDB-220
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95513)
<!-- Reviewable:end -->

tbg committed:
kvstorage: complete RaftReplicaID migration

As of v22.1[^1], we always write the RaftReplicaID when creating a
Replica or updating a snapshot. However, since this is
persisted state that could've originated in older versions and not
updated yet, we couldn't rely on a persisted ReplicaID yet.

This commit adds code to the `(*Store).Start` boot sequence that

- persists a RaftReplicaID for all initialized replicas (using the
  ReplicaID from the descriptor)
- deletes all uninitialized replicas that don't have a RaftReplicaID
  (since we don't know their ReplicaID at this point).

The second item in theory violates Raft invariants, as uninitialized
Replicas are allowed to vote (though they then cannot accept log
entries). So in theory:

- an uninitialized replica casts a decisive vote for a leader
- it restarts
- code in this commit removes the uninited replica (and its vote)
- delayed MsgVote from another leader arrives
- it casts another vote for the same term for a dueling leader
- now there are two leaders in the same term.

The above in addition presupposes that the two leaders cannot
communicate with each other. Also, even if that is the case, since the
two leaders cannot append to the uninitialized replica (it doesn't
accept entries), we also need additional voters to return at the exact
right time.

Since an uninitialized replica without RaftReplicaID in is necessarily
at least one release old, this is exceedingly unlikely and we will
live with this theoretical risk.

This commit also introduces a few assertions that make sure that
we don't have overlapping initialized replicas (which would be
detected at Store.Start time otherwise while inserting in the
btree, but it's nice to catch this earlier) or duplicate
RangeIDs.

[^1]: https://github.com/cockroachdb/cockroach/pull/75761

Epic: CRDB-220
Release note: None

tbg commented:
Now it's good to review.

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95513)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov and @tbg)

___
*[`-- commits` line 36 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95513#-NMUpUs4BRdF5TQA6s09):*
wasn't it v22.1?

There is also a similar comment in https://github.com/cockroachdb/cockroach/blob/c9f87cf306f7b23835e5d2aa7edf2c2ad38b66dd/pkg/kv/kvserver/store_create_replica.go#L249-L283 which could use some adjustment in light of this PR doing this Replica cleanup.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95513)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov and @sumeerbhola)

___
*[`-- commits` line 36 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/95513#-NMUpUs4BRdF5TQA6s09):*
> wasn't it v22.1?

Yep, thanks.

Updated the comment.


<!-- Sent from Reviewable.io -->

pavelkalinnikov commented on pkg/kv/kvserver/kvstorage/init.go:
```diff
@@ -347,6 +375,20 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
 					ident.StoreID, desc)
 			}
 
+			sp := desc.RSpan().AsRawSpanWithNoLocals()
```
Could you add a comment above saying what this block does? It looks to me like it checks that all range descriptors are non-overlapping.

Do you need the interval tree here? If `IterateRangeDescriptorsFromDisk` returns descriptors ordered by the first key (and the `INVARIANT` also says that begin/end keys are verified to match the key at which the descriptor is located), you only need to check that `current.Key` >= `prev.EndKey`. You can also double-check that `current.Key` > `prev.Key` (unless you trust the iteration func, in which case it's sufficient to make a comment).

Side: this seems out of place in this commit/PR, would prefer to have reviewed it separately.

tbg replied:
I brought it back, since it's a one-liner now. I had it in a separate commit but the whitespace changes made it more trouble than is worth it.

pavelkalinnikov commented on pkg/kv/kvserver/kvstorage/init.go:
```diff
@@ -356,37 +398,103 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
+	// In this method, we migrate to always satisfy the invariant:
+	//
+	// 1. for initialized replicas, use ReplicaID from RangeDescriptor as RaftReplicaID
+	// 2. remove uninitialized replicas lacking a RaftReplicaID.
```
(2) doesn't look like invariant. Did you mean to say that "uninitialized replicas don't have a stored RaftReplicaID"? This isn't yet true either, because `getOrCreateReplica` still writes one asap.

Should we hold off deleting uninit replica IDs from storage until (or after) a PR that would stop writing them too? Otherwise the code will be in a state where we always write them, and remove on restarts, which is a bit wasteful.

pavelkalinnikov replied:
I misunderstood the comment. (1) and (2) describe migration actions, which are needed to arrive at the invariant above. I might be wrong in my interpretation of what this does, will take a fresh look tomorrow.

pavelkalinnikov replied:
Ok, so we're looking at 3 KVs of a `Replica`: `RaftReplicaID`, `HardState`, and `RangeDescriptor`.

<table>
<tr>
	<td> RaftReplicaID
	<td> HardState
	<td> RangeDescriptor
	<td> What?
<tr>
	<td> no
	<td> no
	<td> no
	<td> nothing
<tr>
	<td> no
	<td> no
	<td> yes
	<td> -
<tr>
	<td> no
	<td> yes
	<td> no
	<td> uninitialized (v <= 22.1)
<tr>
	<td> no
	<td> yes
	<td> yes
	<td> initialized (v <= 22.1)
<tr>
	<td> yes
	<td> no
	<td> no
	<td> -
<tr>
	<td> yes
	<td> no
	<td> yes
	<td> -
<tr>
	<td> yes
	<td> yes
	<td> no
	<td> uninitialized (v >= 22.2)
<tr>
	<td> yes
	<td> yes
	<td> yes
	<td> initialized
</table>

Steps 1 and 2 in this comment say that we eliminate the "RaftReplicaID: no" half of this table. The way step 1 and 2 do so is different though, because for uninit replica we don't know the Replica ID; if we did we would write it too; but instead we remove such replicas.

Makes sense.

pavelkalinnikov replied:
It's a bit hard to follow all the map maths below. I will try now, but I think this could be simpler if we just implemented something like a table that a drew above:

```golang
type rangeState struct {
	replicaID roachpb.ReplicaID
	hardState raftpb.HardState
	rangeDesc roachpb.RangeDescriptor
}

var m map[roachpb.RangeID]rangeState

read all RaftReplicaID
insert into m[rangeID].replicaID (init a new rangeState if it does not exist)

read all HardState
insert into m[rangeID].hardState (init a new rangeState if it does not exist)

read all RangeDescriptor
insert into m[rangeID].rangeDesc (init a new rangeState if it does not exist)

for rangeID, state := range m {
	checkInvariants(rangeID, state) // e.g. assert that we don't see one of the impossible "-" combinations
	writeMigration(rangeID, state)
}
```

tbg replied:
Good idea, this simplified it a lot. I'll ping you when the review is ready.

pavelkalinnikov commented on pkg/kv/kvserver/kvstorage/init.go:
```diff
@@ -356,37 +398,103 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
+
+	}
+
+	for id := range initM {
```
Not reviewed today below this line. Will continue tomorrow, but also would like to see if this PR can be split: (1) complete backfilling ReplicaIDs, (2) some time later also look at removing ReplicaIDs that we don't need (rationale in the prev. comment).

tbg replied:
I rewrote according to your suggestion and I think the result is orders of magnitude clearer. I tend to think splitting it at this point is more busy work for me than benefit for you but let me know (once I've pinged) and I can still tease it apart after.

pavelkalinnikov commented on pkg/kv/kvserver/kvstorage/init.go:
```diff
@@ -356,37 +398,103 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
-			}
-			log.Eventf(ctx, "backfilled replicaID for %s", id)
+	for id := range allM {
+		if id.ReplicaID != 0 {
+			// RaftReplicaID was present, no need to backfill.
+			continue
 		}
-		// `allM` will be our map of uninitialized replicas.
+
+	}
```
This loop looks like a no-op? Or maybe it's some indentation that confused me.

tbg commented:
Ok, RFAL.

tbg commented:
https://teamcity.cockroachdb.com/repository/download/Cockroach_Ci_Tests_Testrace/8563775:id/bazel-testlogs/pkg/storage/storage_test/test.log

```
panic: concurrent write operations detected on file [recovered]
	panic: concurrent write operations detected on file

goroutine 12121 [running]:
github.com/cockroachdb/pebble.(*DB).runCompaction.func1()
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:2258 +0x23d
panic({0x27ac820, 0x31d3a20})
	GOROOT/src/runtime/panic.go:890 +0x262
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).timeDiskOp(0xc000584910, 0x2, 0xc000988558)
	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:262 +0x1a5
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).Sync(0xc000584910)
	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:219 +0x69
github.com/cockroachdb/pebble/vfs.(*enospcFile).Sync(0xc000138648)
	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_full.go:391 +0x6d
github.com/cockroachdb/pebble.(*DB).runCompaction(0xc00090a500, 0x7d, 0xc000539200)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:2836 +0x312f
github.com/cockroachdb/pebble.(*DB).flush1(0xc00090a500)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1685 +0x61d
github.com/cockroachdb/pebble.(*DB).flush.func1({0x31f15b8, 0xc000a0f2c0})
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1624 +0x137
runtime/pprof.Do({0x31f1548, 0xc0000600a8}, {{0xc0000acf60?, 0xc00090a500?, 0x0?}}, 0xc00073a7a0)
	GOROOT/src/runtime/pprof/runtime.go:40 +0x123
github.com/cockroachdb/pebble.(*DB).flush(0xc00090a500)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1617 +0x92
created by github.com/cockroachdb/pebble.(*DB).maybeScheduleFlush
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1533 +0x186
I230202 22:16:32.090098 1 (gostd) testmain.go:626  [T1] 1  Test //pkg/storage:storage_test exited with error code 2
```

knz commented:
@RaduBerinde requested that we file these concurrency issues inside pebble as issues on the pebble repo.

tbg commented:
Filed https://github.com/cockroachdb/pebble/issues/2301

pavelkalinnikov commented on pkg/kv/kvserver/kvstorage/testdata/init:
```diff
@@ -0,0 +1,39 @@
+# Uninitialized deprecated replica: doesn't have a RaftReplicaID, it's just a
+# HardState. We expect this to be removed by load-and-reconcile.
+new-replica range-id=5
```
r5 doesn't show up in any way in the output below, except in comments that reason about what happened to it. Is there a way to surface it in the output? Or is the absence of r5 in the output the intended signal?

tbg replied:
It does show up:

<img width="652" alt="image" src="https://user-images.githubusercontent.com/5076964/216624048-eb226e16-c994-489b-a201-a9e88f26d761.png">

Basically it was tossed out.

pavelkalinnikov commented on pkg/kv/kvserver/kvstorage/init.go:
```diff
@@ -291,46 +293,93 @@ func IterateRangeDescriptorsFromDisk(
-type EngineReplicas struct {
-	Uninitialized map[storage.FullReplicaID]struct{}
-	Initialized   map[storage.FullReplicaID]*roachpb.RangeDescriptor
+// LoadedReplicas represents the Replicas present on a storage engine.
+type LoadedReplicas struct {
+	Uninitialized ReplicaMap
+	Initialized   ReplicaMap
 }
```
IMO, this type is not needed, one can return `ReplicaMap` directly and sift through (un)inited replicas if needed.

tbg replied:
Done!

pavelkalinnikov commented on pkg/kv/kvserver/kvstorage/init.go:
```diff
@@ -291,46 +293,93 @@ func IterateRangeDescriptorsFromDisk(
+type Replica struct {
+	RangeID   roachpb.RangeID
+	ReplicaID roachpb.ReplicaID
+	Desc      *roachpb.RangeDescriptor // nil for uninitialized Replica
```
Would a default-constructed `RangeDescriptor` suffice (treat `Desc.RangeID == 0` as nil), or you are avoiding some overhead here by making it a pointer?

tbg replied:
We could use the zero value here, but what's the upshot? I know there's the Go koan of "making the zero value useful" but a `RangeDescriptor` can not be compared to its zero value (since it contains a slice) and besides it's much more noisy than just checking if a pointer is nil. For better or worse a nil pointer is the closest we got to a rust `Option` type (maybe generics help 👿 ).

pavelkalinnikov commented on pkg/kv/kvserver/kvstorage/init.go:
```diff
@@ -291,46 +293,93 @@ func IterateRangeDescriptorsFromDisk(
-	// TODO(sep-raft-log): if there is any other data that we mandate is present here
-	// (like a HardState), validate that here.
+// SortedSlice returns the contents of the map as a RangeID-sorted slice.
+func (m ReplicaMap) SortedSlice() []Replica {
```
nit: I'd name it `Sorted()`. The "slice" bit is obvious.

pavelkalinnikov commented on pkg/kv/kvserver/kvstorage/init.go:
```diff
@@ -291,46 +293,93 @@ func IterateRangeDescriptorsFromDisk(
-	// (like a HardState), validate that here.
+// SortedSlice returns the contents of the map as a RangeID-sorted slice.
+func (m ReplicaMap) SortedSlice() []Replica {
+	var sl []Replica
```
Pre-allocate for `len(m)`?

pavelkalinnikov commented on pkg/kv/kvserver/kvstorage/init.go:
```diff
@@ -291,46 +293,93 @@ func IterateRangeDescriptorsFromDisk(
+	return nil
+}
+
+func (m ReplicaMap) setDescReplicaID(rangeID roachpb.RangeID, descReplicaID roachpb.ReplicaID) {
+	ent := m.getOrMake(rangeID)
+	ent.descReplicaID = descReplicaID
+	m[rangeID] = ent
+}
```
Why is this needed? Could we check the invariant using `Desc -> ... -> ReplicaID` directly instead of pre-computing / denormalizing it when loading?

tbg replied:
Good idea, doing this.

pavelkalinnikov commented on pkg/kv/kvserver/kvstorage/init.go:
```diff
@@ -347,47 +396,103 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
 					ident.StoreID, desc)
 			}
```
I think we could push the invariant check down where you check other invariants. This would allow removing `Replica.descReplicaID` and `setDescReplicaID`.

pavelkalinnikov commented on pkg/kv/kvserver/kvstorage/init.go:
```diff
@@ -347,47 +396,103 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
+	//
+	// This invariant is true for replicas created in 22.1. Without further action, it
+	// would be violated for clusters that originated before 22.1. In this method, we
+	// backfill the ReplicaID (for initialized replicas) and we remove uninitialized
+	// replicas (see below for rationale).
```
Removing all uninitialized replicas? Or the ones that don't have a `RaftReplicaID`? Worth clarifying here and in the commit description.

tbg replied:
Done.

tbg replied:
(in comment, commit, PR)

pavelkalinnikov commented on pkg/kv/kvserver/kvstorage/init.go:
```diff
@@ -347,47 +396,103 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
 			}
-			if err := logstore.NewStateLoader(id.RangeID).SetRaftReplicaID(ctx, eng, id.ReplicaID); err != nil {
-				return nil, errors.Wrapf(err, "backfilling replicaID for r%d", id.RangeID)
+			s.setReplicaID(repl.RangeID, repl.descReplicaID)
```
This `set` modifies Go `map` under the hood. I wonder how that impacts the iteration on it, do you know? Maybe a more transparent approach would be:

```golang
repls := s.SortedSlice()
for i, repl := range repls {
	repls[i].foo = bar
}
```

UPD: aah, we're iterating the slice, not the map.

pavelkalinnikov replied:
Since we've sorted it already either way, I would return the sorted slice from the func, so that the caller doesn't need to sort it again.

tbg replied:
Also, we're updating an existing element in the map, so it's fine. If you're adding elements or removing while iterating, the element may or may not show up.

pavelkalinnikov commented on pkg/kv/kvserver/kvstorage/init.go:
```diff
@@ -347,47 +396,103 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
-			if err := logstore.NewStateLoader(id.RangeID).SetRaftReplicaID(ctx, eng, id.ReplicaID); err != nil {
-				return nil, errors.Wrapf(err, "backfilling replicaID for r%d", id.RangeID)
+			s.setReplicaID(repl.RangeID, repl.descReplicaID)
+			log.Eventf(ctx, "backfilled replicaID for initialized replica %s", s.getOrMake(repl.RangeID).ID())
```
nit: `getOrMake` seems unnecessary here, could construct `ID` directly.

pavelkalinnikov commented on pkg/kv/kvserver/kvstorage/init.go:
```diff
@@ -347,47 +396,103 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
+			uninit[rangeID] = state
 		}
-		// `allM` will be our map of uninitialized replicas.
-		//
-		// A replica is "uninitialized" if it's not in initM (i.e. is at log position
-		// zero and has no visible RangeDescriptor).
-		delete(allM, id)
 	}
 
-	return &EngineReplicas{
-		Initialized:   initM,
-		Uninitialized: allM, // NB: init'ed ones were deleted earlier
-	}, nil
+	return LoadedReplicas{Initialized: init, Uninitialized: uninit}, nil
```
`return s`?

pavelkalinnikov commented on pkg/kv/kvserver/store_create_replica.go:
```diff
@@ -245,42 +245,19 @@ func (s *Store) tryGetOrCreateReplica(
 	//   this newer replica is harmless since it just limits the votes for
```
unrelated drive-by question: Is "earlier replica" possible here? Why only "same" or "later"?

tbg replied:
Earlier might be possible. I think there is a hole here, the tombstone and the RaftReplicaID must really work in unison - if RaftReplicaID is 10 then a tombstone of 5 should be treated as a tombstone of 9.

Let's discuss on Slack, I'd like to get this PR in :-)

pav-kv reviewed (approved):
Looking much clearer now, thanks for implementing the suggestion! LGTM in principle, but left some tips/nits.

tbg commented on pkg/kv/kvserver/kvstorage/datadriven_test.go:
```diff
@@ -0,0 +1,208 @@
+				}
+				for id := range rs.Uninitialized {
+					merged = append(merged, id)
+				}
```
Good suggestion, that simplified things.

tbg commented:
Still have to push the updates, please ignore my comment pings until I request.

tbg commented on pkg/kv/kvserver/kvstorage/testdata/init:
```diff
@@ -0,0 +1,41 @@
+r8/80: r8:{c-f} [(n1,s1):80, next=81, gen=0]
+beginning range descriptor iteration
+iterated over 2 keys to find 2 range descriptors (by suffix: map[rdsc:2])
+removed legacy uninitialized replica for r5
```
Here, @pavelkalinnikov

pavelkalinnikov commented on pkg/kv/kvserver/kvstorage/init.go:
```diff
@@ -291,103 +293,197 @@ func IterateRangeDescriptorsFromDisk(
+					return errors.AssertionFailedf("overlapping descriptors %s and %s", lastDesc, desc)
+				}
+				lastDesc = desc
+
+				if err := s.setDesc(desc.RangeID, desc); err != nil {
+					return err
+				}
+				return nil
```
supernit: `return s.setDesc(...)`

pavelkalinnikov commented on pkg/kv/kvserver/kvstorage/init.go:
```diff
@@ -291,103 +293,197 @@ func IterateRangeDescriptorsFromDisk(
+					return err
+				}
+				return nil
+			}); err != nil {
```
supernit: I'd put comma here and `); err != nil` check on the next line to avoid the weird indentation.

pavelkalinnikov commented on pkg/kv/kvserver/kvstorage/init.go:
```diff
@@ -291,103 +293,197 @@ func IterateRangeDescriptorsFromDisk(
+	ent := m.getOrMake(rangeID)
+	ent.ReplicaID = replicaID
+	m[rangeID] = ent
+	return ent
```
nil: dunno, this is a bit weird, can we avoid this API?

I think (idea 1):
- The user of `LoadAndReconcileReplicas` doesn't need the map, so it can return `[]Replica` directly.
- Since we're sorting anyway (for determinism), let it return a sorted `[]Replica`.
- Then, the tests don't need to `Sorted()` the map again.
- Also, `ReplicaMap` and all its methods (incl. `Sorted`) can be unexported.
- Inside `LoadAndReconcileReplicas` in the place where we start overriding things in the map, we should just convert to sorted slice, and modify slice elements directly.

Idea 2: If you really want to expose the map-ness to the outside (e.g. the user of this needs to search by RangeID, which though it currently doesn't), I could also imagine a data structure like this:

```golang
type ReplicaMap struct {
	index map[roachpb.RangeID]int
	repls []Replica
}

func (m *ReplicaMap) getOrMake(rangeID roachpb.RangeID) int { // or *Replica
	if index, found := m[rangeID]; found {
		return index
	}
	m.repls = append(m.repls, Replica{RangeID: rangeID})
	return len(m.repls) - 1
}

func (m *ReplicaMap) Sort() { // or stateless Sorted() like right now
	// sort m.repls
	// reindex m.index using m[].RangeID as key
}

// then all the accessors are not needed, can just `getOrMake` and write
// to the relevant returned *Replica (or repls[index]) directly, like this:
// 
m.getOrMake(rangeID).hardState = hs
// or
m.repls[m.getOrMake(rangeID)].hardState = hs
```

Some combination of idea 1 and 2 would be nice.

tbg replied:
Went with 1), the yak has been shaved 👑 
I like the result better.

pav-kv reviewed (approved):
Some final bits and bobs.

pavelkalinnikov commented on pkg/kv/kvserver/kvstorage/init.go:
```diff
@@ -291,103 +293,198 @@ func IterateRangeDescriptorsFromDisk(
 			}
-			log.Eventf(ctx, "backfilled replicaID for %s", id)
+			log.Eventf(ctx, "removed legacy uninitialized replica for r%s", repl.RangeID)
+			sl = append(sl[:idx], sl[idx+1:]...)
```
To avoid the O(N * M) complexity here, in a case when N out of M replicas are legacy uninited replicas, could instead rewrite one `Replica` in the slice at a time:

```golang
back := 0
for idx, repl := range sl {
	...
	if !(remove condition) {
		sl[back] = repl
		back++
	}
}
return sl[:back], nil
```

May require a bit of shuffling/reordering to work around the `continue`.

tbg commented:
bors r=pavelkalinnikov
